### DISCUSSION TITLE
Compatibility Next 15 and React 19

### DISCUSSION THREAD

**Discussion by tonyherbert** [react-page/react-page#1331] [Q&A]
Hi everyone,

I am starting a project based on next 15 and react 19 and I see that apparently react page is not compatible because of Dnd which conflicts with react 19.

Can anyone tell me if this is in the future projects?
I really like React page!

thank you guys !
<img width="949" alt="image" src="https://github.com/user-attachments/assets/d54d3992-8aeb-48d9-a7aa-d7ea4937b094">

**Comment by fcreme** [comment]
I went and actually tested this, because the DnD peer-dependency story turns out to be a red herring. Setup: Next 15.5.25, React 19.3.0, `@react-page/editor` 5.4.6 (current latest), `@react-page/plugins-slate` + `plugins-image` 5.4.6, npm 11.

Short version: it doesn't work out of the box, but two of the three problems are fixable in your own `package.json`, and the third is a three-line fix upstream.

### 1. react-dnd's peer range is not the conflict

Both react-dnd 15 and 16 declare `"react": ">= 16.14"`, which already allows React 19. It never produces an `ERESOLVE` against React 19.

The two packages that actually cap at React 18 are:

* **`react-redux@7.2.9`** → `^16.8.3 || ^17 || ^18`. npm prints `ERESOLVE overriding peer dependency` and carries on. It does work, because `react-dom@19` still exports `unstable_batchedUpdates`, which is what v7 needs.
* **`uniforms@3.10.2`** (via `uniforms-bridge-json-schema`) → `^16.8.0 || ^17 || ^18`. This one is sneakier: npm doesn't error, it quietly installs a **second copy of React 18.3.1** at `node_modules/uniforms-bridge-json-schema/node_modules/react`. Two Reacts in one tree.

### 2. What actually fails the build

```
./node_modules/react-dnd/dist/esm/core/DndProvider.mjs
Module not found: Package path ./jsx-runtime.js is not exported from package .../node_modules/react
```

react-dnd **15**'s ESM build does `import { jsx as _jsx } from "react/jsx-runtime.js"` — with the `.js`. React's `exports` map only exposes `./jsx-runtime`, and Next 15's webpack enforces the exports map strictly. react-dnd **16.0.1 removed that import entirely** (no `react/jsx-runtime.js` reference left anywhere in the package), and its peer range is identical to 15's, so it's a drop-in.

You'll also need `@emotion/react` / `@emotion/styled` — MUI 5 needs them, but they're *optional* peers, so npm won't install them for you.

### 3. The real React 19 blocker is react-draggable

`@react-page/editor` depends on `"react-draggable": "4.4.5"` — an exact pin, no caret — and uses it in `packages/editor/src/core/components/Row/ResizableRowCell.tsx` **without passing `nodeRef`**. react-draggable's internal lookup is:

```js
findDOMNode() {
  return this.props?.nodeRef?.current ?? ReactDOM.findDOMNode(this);
}
```

and React 19 removed `ReactDOM.findDOMNode`. So as soon as a row has two or more columns in edit mode, `showResizeHandle` flips true, `<Draggable>` mounts, and the whole page dies:

```
Uncaught TypeError: ReactDOM.findDOMNode is not a function
```

In Next that surfaces as the generic *"Application error: a client-side exception has occurred"*. Worth stressing: **bumping react-draggable does not fix this** — 4.7.1 still falls back to `ReactDOM.findDOMNode` when no `nodeRef` is given. Passing `nodeRef` is the only fix.

### A working setup

`package.json`:

```jsonc
{
  "dependencies": {
    "@react-page/editor": "5.4.6",
    "@emotion/react": "^11.14.0",
    "@emotion/styled": "^11.14.0",
    "next": "^15.0.0",
    "react": "^19.0.0",
    "react-dom": "^19.0.0"
  },
  "overrides": {
    "react-dnd": "16.0.1",
    "react-dnd-html5-backend": "16.0.1",
    "uniforms": { "react": "$react" },
    "uniforms-bridge-json-schema": { "react": "$react" }
  }
}
```

(yarn: `resolutions`; pnpm: `pnpm.overrides` plus `peerDependencyRules`.)

Import it client-side only, as before:

```jsx
'use client';
import dynamic from 'next/dynamic';
const Editor = dynamic(() => import('./MyEditor'), { ssr: false });
```

Then patch `ResizableRowCell` (patch-package, or `pnpm patch`). The upstream change is three lines:

```diff
 const ResizableRowCell: React.FC<...> = ({ nodeId, rowWidth, ... }) => {
+  const handleRef = React.useRef<HTMLDivElement>(null);
   ...
-    <Draggable bounds={...} position={...} axis="x" onDrag={...} grid={...}>
-      <div className="resize-handle" style={...} onClick={...} />
+    <Draggable nodeRef={handleRef} bounds={...} position={...} axis="x" onDrag={...} grid={...}>
+      <div ref={handleRef} className="resize-handle" style={...} onClick={...} />
     </Draggable>
```

With those three changes: clean `next build`, editor mounts under `reactStrictMode: true`, slate and image cells render, the uniforms-based plugin settings form works, and dragging the column handle resizes correctly (6/6 → 7/5 → 8/4 → 9/3 → 10/2) with zero console errors or warnings.

### On "is this planned?"

Setting expectations honestly: the only commit between April 2023 and July 2026 was a README edit. The July 2026 burst that produced 5.4.6 was dependency-floor/security and yarn-4 build tooling work, not React 19 support, and there's no open React 19 issue or PR.

@macrozone — happy to open a PR for the `nodeRef` fix and the react-dnd 15 → 16 bump if that's useful. Both are small and stay compatible with React 16.14/17/18, so they wouldn't be breaking changes.
